### PR TITLE
[SPARK-23014][SS] Fully remove V1 memory sink.

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
## What changes were proposed in this pull request?

There is a MemorySink v2 already so v1 can be removed. In this PR I've removed it completely.
What this PR contains:
* V1 memory sink removal
* V2 memory sink renamed to become the only implementation
* Since DSv2 sends exceptions in a chained format (linking them with cause field) I've made python side compliant
* Adapted all the tests

## How was this patch tested?

Existing unit tests.

### CODE REVIEW

**Comment by gaborgsomogyi** [comment]
Just for the sake of understanding why python adaptation needed.
With V1 memory sink the following exception arrived:
```
Job aborted due to stage failure: Task 0 in stage 1.0 failed 1 times, most recent failure: Lost task 0.0 in stage 1.0 (TID 1, localhost, executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/Users/gaborsomogyi/spark/python/lib/pyspark.zip/pyspark/worker.py", line 428, in main
    process()
  File "/Users/gaborsomogyi/spark/python/lib/pyspark.zip/pyspark/worker.py", line 423, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "/Users/gaborsomogyi/spark/python/pyspark/serializers.py", line 457, in dump_stream
    self.serializer.dump_stream(self._batched(iterator), stream)
  File "/Users/gaborsomogyi/spark/python/pyspark/serializers.py", line 141, in dump_stream
    for obj in iterator:
  File "/Users/gaborsomogyi/spark/python/pyspark/serializers.py", line 446, in _batched
    for item in iterator:
  File "<string>", line 1, in <lambda>
  File "/Users/gaborsomogyi/spark/python/lib/pyspark.zip/pyspark/worker.py", line 86, in <lambda>
    return lambda *a: f(*a)
  File "/Users/gaborsomogyi/spark/python/pyspark/util.py", line 99, in wrapper
    return f(*args, **kwargs)
  File "/Users/gaborsomogyi/spark/python/pyspark/sql/tests/test_streaming.py", line 217, in <lambda>
    bad_udf = udf(lambda x: 1 / 0)
ZeroDivisionError: integer division or modulo by zero

    at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:453)
    at org.apache.spark.sql.execution.python.PythonUDFRunner$$anon$2.read(PythonUDFRunner.scala:81)
    at org.apache.spark.sql.execution.python.PythonUDFRunner$$anon$2.read(PythonUDFRunner.scala:64)
    at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)
    at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
    at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:489)
    at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
    at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
    at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage2.processNext(generated.java:25)
    at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
    at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:748)
    at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:255)
    at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:852)
    at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:852)
    at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:327)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:291)
    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
    at org.apache.spark.scheduler.Task.run(Task.scala:121)
    at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:425)
    at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1349)
    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:428)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)

Driver stacktrace:
=== Streaming Query ===
Identifier: this_query [id = fcda74c8-05e0-4c74-ac45-3444f3e50c63, runId = 7534a374-5d15-4986-bae1-8fa0664c647b]
Current Committed Offsets: {}
Current Available Offsets: {FileStreamSource[file:/Users/gaborsomogyi/spark/python/test_support/sql/streaming]: {"logOffset":0}}

Current State: ACTIVE
Thread State: RUNNABLE

Logical Plan:
Project [<lambda>(value#4) AS <lambda>(value)#17]
+- StreamingExecutionRelation FileStreamSource[file:/Users/gaborsomogyi/spark/python/test_support/sql/streaming], [value#4]
```
With V2 memory sink the following exception arrived (here additional exceptions are linked to the top level one, for instance `ZeroDivisionError`):
```
Writing job aborted.
=== Streaming Query ===
Identifier: this_query [id = d37ca2ea-edb7-4670-b4cc-75b0d704496e, runId = 587d62da-8355-4f66-a9d8-708c154ef953]
Current Committed Offsets: {}
Current Available Offsets: {FileStreamSource[file:/Users/gaborsomogyi/spark/python/test_support/sql/streaming]: {"logOffset":0}}

Current State: ACTIVE
Thread State: RUNNABLE

Logical Plan:
WriteToMicroBatchDataSource org.apache.spark.sql.execution.streaming.sources.MemoryStreamingWrite@1a3691ef
+- Project [<lambda>(value#18) AS <lambda>(value)#32]
   +- StreamingExecutionRelation FileStreamSource[file:/Users/gaborsomogyi/spark/python/test_support/sql/streaming], [value#18]
```

**Comment by gaborgsomogyi** [comment]
cc @jose-torres since you've started this originally and @HyukjinKwon because of the python change.

**Comment by gaborgsomogyi** [comment]
Not much movement cc @vanzin

File: core/src/main/scala/org/apache/spark/TestUtils.scala [hunk 1]
```diff
@@ -192,6 +192,20 @@ private[spark] object TestUtils {
     assert(listener.numSpilledStages == 0, s"expected $identifier to not spill, but did")
   }
 
+  /**
+   * Asserts if exception message doesn't contain the message. Please note this checks all
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/TestUtils.scala]: "that exception message contains"

At least that's what the code seems to be doing.

File: core/src/main/scala/org/apache/spark/TestUtils.scala [hunk 2]
```diff
@@ -192,6 +192,20 @@ private[spark] object TestUtils {
...
+   * exceptions in the tree.
+   */
+  def assertExceptionMsg(exception: Throwable, msg: String): Unit = {
+    var e = exception
+    var contains = e.getMessage.contains(msg)
+    while (e.getCause != null && contains == false) {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/TestUtils.scala]: `&& !contains`

File: mllib/src/test/scala/org/apache/spark/ml/recommendation/ALSSuite.scala [hunk 3]
```diff
@@ -701,7 +701,7 @@ class ALSSuite extends MLTest with DefaultReadWriteTest with Logging {
         }.getMessage.contains(msg))
         assert(intercept[StreamingQueryException] {
           testTransformer[A](dataFrame, model, "prediction") { _ => }
-        }.getMessage.contains(msg))
+        }.getCause.getCause.getMessage.contains(msg))
```
Comment (vanzin) [inline mllib/src/test/scala/org/apache/spark/ml/recommendation/ALSSuite.scala]: Could you use your new helper method here? Calling `getCause` like that feels a bit weird.

File: python/pyspark/sql/streaming.py [hunk 4]
```diff
@@ -186,7 +186,7 @@ def exception(self):
             je = self._jsq.exception().get()
             msg = je.toString().split(': ', 1)[1]  # Drop the Java StreamingQueryException type info
             stackTrace = '\n\t at '.join(map(lambda x: x.toString(), je.getStackTrace()))
-            return StreamingQueryException(msg, stackTrace)
+            return StreamingQueryException(msg, stackTrace, je.getCause())
```
Comment (vanzin) [inline python/pyspark/sql/streaming.py]: To be super safe, should you check that `je.getCause()` exists?

File: python/pyspark/sql/tests/test_streaming.py [hunk 5]
```diff
@@ -225,11 +225,19 @@ def test_stream_exception(self):
...
         finally:
             sq.stop()
         self.assertTrue(type(sq.exception()) is StreamingQueryException)
-        self.assertTrue("ZeroDivisionError" in sq.exception().desc)
+        self._assert_exception_tree_contains_msg(sq.exception(), "ZeroDivisionError")
+
+    def _assert_exception_tree_contains_msg(self, exception, msg):
+        e = exception
+        contains = msg in e.desc
+        while e.cause is not None and contains is False:
```
Comment (vanzin) [inline python/pyspark/sql/tests/test_streaming.py]: `not contains`

File: python/pyspark/sql/tests/test_streaming.py [hunk 6]
```diff
@@ -225,11 +225,19 @@ def test_stream_exception(self):
...
+            e = e.cause
+            contains = msg in e.desc
+        self.assertTrue(contains)
```
Comment (vanzin) [inline python/pyspark/sql/tests/test_streaming.py]: Add a message to the assert?

File: python/pyspark/sql/utils.py [hunk 7]
```diff
@@ -57,27 +58,41 @@ class QueryExecutionException(CapturedException):
...
-                raise ParseException(s.split(': ', 1)[1], stackTrace)
-            if s.startswith('org.apache.spark.sql.streaming.StreamingQueryException: '):
-                raise StreamingQueryException(s.split(': ', 1)[1], stackTrace)
-            if s.startswith('org.apache.spark.sql.execution.QueryExecutionException: '):
-                raise QueryExecutionException(s.split(': ', 1)[1], stackTrace)
-            if s.startswith('java.lang.IllegalArgumentException: '):
-                raise IllegalArgumentException(s.split(': ', 1)[1], stackTrace)
-            raise
+            converted = convert_exception(e.java_exception)
+            if type(converted) is not UnknownException:
```
Comment (vanzin) [inline python/pyspark/sql/utils.py]: not sure what's the usual style in pyspark but I've generally used `isinstance(foo, Type)` for this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/memory.scala [hunk 8]
```diff
@@ -128,14 +128,14 @@ class MemorySinkV2 extends SupportsStreamingWrite with MemorySinkBase with Loggi
...
-  override def toString(): String = "MemorySinkV2"
+  override def toString(): String = "MemorySink"
 }
 
 case class MemoryWriterCommitMessage(partition: Int, data: Seq[Row])
   extends WriterCommitMessage {}
 
 class MemoryStreamingWrite(
-    val sink: MemorySinkV2, schema: StructType, needTruncate: Boolean)
+                            val sink: MemorySink, schema: StructType, needTruncate: Boolean)
```
Comment (vanzin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/memory.scala]: indentation

**Review by vanzin — COMMENTED** [review]
Just minor things.

File: core/src/main/scala/org/apache/spark/TestUtils.scala (hunk 2 above)
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/TestUtils.scala]: Fixed.

File: python/pyspark/sql/streaming.py (hunk 4 above)
Comment (gaborgsomogyi) [reply python/pyspark/sql/streaming.py]: Cause check (not None) is inside `class CapturedException`. Do you suggest something different here?

File: python/pyspark/sql/tests/test_streaming.py (hunk 5 above)
Comment (gaborgsomogyi) [reply python/pyspark/sql/tests/test_streaming.py]: Fixed.

File: python/pyspark/sql/tests/test_streaming.py (hunk 6 above)
Comment (gaborgsomogyi) [reply python/pyspark/sql/tests/test_streaming.py]: Message added.

File: python/pyspark/sql/utils.py (hunk 7 above)
Comment (gaborgsomogyi) [reply python/pyspark/sql/utils.py]: Yeah, I think it's more descriptive so changed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/memory.scala (hunk 8 above)
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/memory.scala]: Fixed.

File: core/src/main/scala/org/apache/spark/TestUtils.scala (hunk 1 above)
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/TestUtils.scala]: I was hesitating what you mean here but now I see the point.
`Asserts if` == `throws exception if` was my intention but reading this back as an english sentence is just bad. Fixed.

File: python/pyspark/sql/streaming.py (hunk 4 above)
Comment (vanzin) [reply python/pyspark/sql/streaming.py]: I mean if `je.getCause()` is `None`, shouldn't you be using `je` so that the error is not lost?

**Comment by vanzin** [comment]
Merging to master.

**Pull request closed without merging**